erpnext/setup/doctype/transaction_deletion_tool/transaction_deletion_tool.py

```diff
@@ -0,0 +1,144 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2021, Frappe Technologies Pvt. Ltd. and contributors
+# For license information, please see license.txt
+
+from __future__ import unicode_literals
+from frappe.utils import cint
+import frappe
+from frappe.model.document import Document
+from frappe import _
+from frappe.desk.notifications import clear_notifications
+
+class TransactionDeletionTool(Document):
+	def validate(self):
+		frappe.only_for('System Manager')
+		company_obj = frappe.get_doc('Company', self.company)
+		if frappe.session.user != company_obj.owner and frappe.session.user != 'Administrator':
+			frappe.throw(_('Transactions can only be deleted by the creator of the Company or the administrator.'), 
+				frappe.PermissionError)
+		doctypes_to_be_ignored_list = get_doctypes_to_be_ignored()
+		for doctype in self.doctypes_to_be_ignored:
+			if doctype.doctype_name not in doctypes_to_be_ignored_list:
+				frappe.throw(_("DocTypes should not be added manually to the 'DocTypes That Won't Be Affected' table."))
+
+	def before_submit(self):
+		if not self.doctypes_to_be_ignored:
+			self.populate_doctypes_to_be_ignored_table()
+
+		self.delete_bins()
+		self.delete_lead_addresses()
+		
+		company_obj = frappe.get_doc('Company', self.company)
+		# reset company values
+		company_obj.total_monthly_sales = 0
+		company_obj.sales_monthly_history = None
+		company_obj.save()
+		# Clear notification counts
+		clear_notifications()
+
+		singles = frappe.get_all('DocType', filters = {'issingle': 1}, pluck = 'name')
+		tables = frappe.get_all('DocType', filters = {'istable': 1}, pluck = 'name')
+		doctypes_to_be_ignored_list = singles + get_doctypes_to_be_ignored()
```

Comment:
`get_doctypes_to_be_ignored` returns all the hard-coded doctypes. If a user removes a doctype from that hard-coded list, it will not be accounted for as all the doctypes are fetched here.

You should filter out the doctype names from `self.doctypes_to_be_ignored` and append to `doctypes_to_be_ignored_list`

Reply:
but isn't `self.doctypes_to_be_ignored` populated using `get_doctypes_to_be_ignored()` as well? So any changes to the hard-coded list would reflect on both, right? Unless the hard-coded list is changed after the table gets populated

Reply:
Consider if user removed 'Customer' from the excluded doctypes table after saving the document, then it won't be ignored as the ignored list is not derived from "self.doctypes_to_be_ignored"

Reply:
Ohh right, that makes sense.